src/lib.rs

```diff
@@ -55,21 +55,33 @@ where
     where
         's: 'cx,
     {
-        let curr = self
-            .status
-            .current
-            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
-
-        if curr > self.status.limit {
-            self.status
+        loop {
+            let curr = self
+                .status
                 .current
-                .fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
+                .load(std::sync::atomic::Ordering::Relaxed);
+
+            if curr >= self.status.limit {
```

Comment:
In this implementation, curr will not greater than limit?

Reply:
Yes, since we changed from `modify-check-modify` to `read-check-modify`, both `curr` and barrier (actually effect concurrency) will not greater than limit